game/scripts/vscripts/abilities/oaa_kunkka_tidebringer.lua

```diff
@@ -0,0 +1,236 @@
+  return "particles/units/heroes/hero_kunkka/kunkka_weapon_tidebringer.vpcf"
+end
+]]
+
+--------------------------------------------------------------------------------
+
+function modifier_kunkka_tidebringer_oaa_passive:OnCreated()
+  local parent = self:GetParent()
+  -- Add weapon glow
+  parent:AddNewModifier(parent, self:GetAbility(), "modifier_kunkka_tidebringer_oaa_weapon_effect", {})
+end
+
+function modifier_kunkka_tidebringer_oaa_passive:DeclareFunctions()
+  return {
+    MODIFIER_EVENT_ON_ATTACK_LANDED,
+    MODIFIER_PROPERTY_PREATTACK_BONUS_DAMAGE
+  }
+end
+
+function modifier_kunkka_tidebringer_oaa_passive:GetModifierPreAttack_BonusDamage( event )
+  if self:GetParent() ~= event.attacker then
+    return
+  end
+
+  -- Toggled on
+  local ability = self:GetAbility()
+  if ability:GetAutoCastState() and ability:IsCooldownReady() then
+    return ability:GetTalentSpecialValueFor("damage_bonus")
+  end
+  return 0
+end
+
+function modifier_kunkka_tidebringer_oaa_passive:OnAttackLanded( event )
+  -- Only attacks FROM parent
+  if self:GetParent() ~= event.attacker then
+    return
+  end
+
+  local parent = self:GetParent()
+  local ability = self:GetAbility()
+
+  -- Toggled off
+  if ( not ability:GetAutoCastState() ) or ( not ability:IsCooldownReady() ) then
+    return
+  end
+
+  local cleaveInfo = {
+    startRadius = ability:GetTalentSpecialValueFor("cleave_starting_width"),
+    endRadius = ability:GetTalentSpecialValueFor("cleave_ending_width"),
+    length = ability:GetTalentSpecialValueFor("cleave_distance")
+  }
+  local hitUnits = ability:PerformCleaveOnAttack(
+    event,
+    cleaveInfo,
+    ability:GetTalentSpecialValueFor("cleave_damage") / 100.0,
+    "Hero_Kunkka.Tidebringer.Attack",
+    "Hero_Kunkka.TidebringerDamage",
+    "particles/units/heroes/hero_kunkka/kunkka_spell_tidebringer.vpcf"
+  )
+
+  -- Force cooldown
+  -- Including cooldown reduction which is why we do not use StartCooldown()
+  ability:UseResources(true, true, true)
+  -- Remove weapon glow effect
+  parent:RemoveModifierByName("modifier_kunkka_tidebringer_oaa_weapon_effect")
+  -- Add cooldown timer
+  parent:AddNewModifier(parent, ability, "modifier_kunkka_tidebringer_oaa_cooldown", {
+    duration = ability:GetCooldownTimeRemaining()
+  })
+end
+
+--------------------------------------------------------------------------------
+
+modifier_kunkka_tidebringer_oaa_cooldown = class( ModifierBaseClass )
+
+--------------------------------------------------------------------------------
+
+function modifier_kunkka_tidebringer_oaa_cooldown:IsHidden()
+  return true
+end
+
+function modifier_kunkka_tidebringer_oaa_cooldown:IsPurgable()
+  return false
+end
+
+function modifier_kunkka_tidebringer_oaa_cooldown:IsDebuff()
+  return false
+end
+
+function modifier_kunkka_tidebringer_oaa_cooldown:GetAttributes()
+  return bit.bor(MODIFIER_ATTRIBUTE_IGNORE_INVULNERABLE, MODIFIER_ATTRIBUTE_PERMANENT)
```

Comment:
Seems weird to me to give a modifier the permanent attribute when it's not actually permanent. And again, you probably want `RemoveOnDeath` false.

Reply:
I gave it permanent attribute to not be removed by some weird Dota internal mechanic.